src/irr/core/LRUcache.h

```diff
@@ -0,0 +1,58 @@
+#include <list>
+#include "irr/core/Types.h"
+#include "COpenGLDriver.h"
+
+namespace irr {
+	namespace core {
+
+		template<typename Key, typename Value,typename MapHash,typename MapEquals>
+		class LRUcache {
+			list<Key> stored_keys;
+			unordered_map<Key, Value, MapHash,MapEquals> map;
```

Comment:
use a `list<pair<Key,Value>>`, I use it like that in the `DeferredEventHandler` to implement a queue (although I only need forward linked list there).

However you need to create your own implementation of a Doubly Linked List using a pool address allocator such that every node is not dynamically allocated from the heap (its possible that its just enough to implement a custom allocator on top of `PoolAddressAllocator` and pass that in as a template argument for `list`, but only if `list` calls the allocator to allocate one node at a time).

then `unordered_map<Key,stored_keys::iterator,MapHash,MapEquals>` to be the "skip list"

First you need a `common_peek`:
1) look up the hashmap with key
2) use hashmap's iterator (pointer) to the linked list node

When you `peek`:
1) call `common_peek` to find yout node and return node's value

When you `get`:
1) call `common_peek` to find your node
2) move the node from its current position in the list to the back
3) return node's value

When you `insert` (you assume that user cannot call `insert` twice on the same key):
1) if cache cap has been reached, you call `common_erase` on the front linked list node
2) construct a linked list node out of `Key,Value` and add it to the back of the list
3) record the iterator (pointer) to that node into the `unordered_map`

When you `erase`:
1) get the node by calling `common_peek`
2) call `common_erase` on the found node

In `common_erase`:
1) remove node from the linked list
2) look up the node's "skip link" in the unordered map and remove it

Reply:
Since timestamp is redundant, what is the difference between peek and get?

Reply:
explained on discord